kong/core/router.lua

```diff
@@ -600,8 +601,22 @@ function _M.new(apis)
     end
 
 
-    if api_t.upstream.path then
-      new_uri = api_t.upstream.path .. new_uri
+    local upstream_path = api_t.upstream.path
+    if upstream_path then
+      if new_uri == "/" then
+        new_uri = upstream_path
+      else
+        new_uri = upstream_path .. (sub(upstream_path, -1) == "/" and sub(new_uri, 2) or new_uri)
+      end
+    end
+
+
+    local req_uri_slash = sub(uri,     -1) == "/"
+    local new_uri_slash = sub(new_uri, -1) == "/"
+    if new_uri_slash and not req_uri_slash and new_uri ~= "/" then
```

Comment:
maybe add one final commit with all the style changes, updating the entire file?